index/scorch/snapshot_index_tfr.go

```diff
@@ -181,6 +181,13 @@ func (i *IndexSnapshotTermFieldReader) Count() uint64 {
 func (i *IndexSnapshotTermFieldReader) Close() error {
 	if i.snapshot != nil {
 		atomic.AddUint64(&i.snapshot.parent.stats.TotTermSearchersFinished, uint64(1))
+		if bytes.Equal(i.term, OptimizeTFRConjunctionUnadornedTerm) ||
+			bytes.Equal(i.term, OptimizeTFRDisjunctionUnadornedTerm) {
+			// Reset iterators before recycling unadorned term field readers,
+			// so as to release the roaring.Bitmaps set up by AND-ing or OR-ing
+			// individual bitmaps for GC.
+			i.iterators = i.iterators[:0]
```

Comment:
I'm not sure I follow this part, is it because these bitmaps are on the heap, while in the other case the bitmaps are backed by bytes mmap'd from disk?

I really liked the way you organized the rest of the code into the unadornedTermFieldReader method, as it reduced duplication and the possibility that some new field doesn't get initialized correctly down the road.  But here I don't like that IndexSnapshotTermFieldReader has to know anything about optimizations...

Reply:
I was thinking that if we don't let go of these iterators here, the bitmaps would never be released - but I was wrong. This bit here is unnecessary. Let me update the last patchset.

Reply:
So, hate to go back and forth on this, but now it seems like there is something to this.

It seems that when we recycle the TFR, we don't reset any of the slices inside (dicts, postings, iterators).  Now, at first I thought, maybe that was on purpose, but the TFRs are shared across the entire snapshot, so you can't safely reuse any of those fields, you have to reinitialize them anyway (different field, term or both).  To be clear, the only thing we're reusing is the struct itself, not the contents.

So, that means a no-longer-used struct sits in slice, possibly holding on to references to other things.  Is that a problem?  Let's look closer, the slice of dictionaries does use some memory, but complex to build FST is already being cached inside of zap, and bytes back it are mmap'd, so it's just the size of the Dictionary structs that really counts.  Next, the PostingsLists, again the structs themselves, and the roaring.Bitmap structs would take some memory, but, in the case of normal TFRs the bytes backing the roaring are mmap'd.  But, in the optimization case, these are on the heap.  Finally, the PostingsIterators, another struct, possibly another bitmap, and several more things that are generally mmap'd.

In conclusion, it seems like there is sufficient reason to reset all of these slices before recycling, and in all cases, not just the optimization case.  Does this make sense?

Reply:
In today's code we recycle the TFRs without actually reseting any of the dicts, postings and iterators. When the TFRs are picked up for re-use, we only re-use dicts, but overwrite postings and iterators and at this point we release the references from the previous use.

So reseting the postings and iterators slices shouldn't change anything - and is probably the better thing to do.
As for the dicts slice, if we were to reset them - we'd incur additional overhead of having to set them as well during reuse.

Reply:
Ah you're correct, the dicts slice is reused, so we should not reset that one.

Do you agree that resetting the other slices potentially frees up memory sooner?  It eventually gets freed, but not until someone else reuses that instance and overwrites the pointer...

Reply:
Yes, agreed. Let me push up another patch-set here.

Reply:
One more thing, looks like we use the previous postings entry to decide on the next one on re-use. So it's actually just the iterators we can reset safely.

Reply:
So unfortunately, we're back to the original comment here - it makes sense to clear iterators only for unadorned term field readers :/

Reply:
Well I've misread the code 3 times now, so my opinion isn't very helpful.